db/version_set.h

```diff
@@ -1262,6 +1262,20 @@ class VersionSet {
   Status VerifyFileMetadata(const std::string& fpath,
                             const FileMetaData& meta) const;
 
+  // Note on synchronization:
+  // When opening a DB,
+  // 1. the DB recovery thread will read and write wals_.
+  // After recovery,
+  // 2. only the LogAndApply thread will read and write wals_.
+  //
+  // Before entering DB recovery, DB mutex is acquired,
+  // before LogAndApply, DB mutex is also acquired,
+  // so writes to wals_ by the DB recovery thread happens before
```

Comment:
Note that both recovery and `LogAndApply` can release db mutex in the middle.

I think the real reason is that
- no user thread will call `LogAndApply` concurrently with the thread calling `Recover()` because application cannot write to a `db` before the database is opened and `db` pointer is returned. The thread calling `Recover()` also calls `LogAndApply()`.
- threads calling `LogAndApply` after recovery succeeds and db is opened. These threads are serialized via the `manifest_writers_` queue.

I think it is very likely that some APIs will need read access to the `wals_` which can happen concurrently with `LogAndApply`. For example, `DB::GetSortedWalFiles()` may want to use `wals_` to verify the WAL files. If so, some sort of synchronization will be necessary.

Reply:
During recovery, since the read and write are in the same thread, there is no need to synchronize on wals_.

I updated the comment to be more explicit again.

Reply:
Yes, during recovery, `GetSortedWalFiles` may need to access `wals_` concurrently with `LogAndApply`.

For `GetSortedWalFiles`, since it's in the same recovery thread, the `wals_` written during manifest recovery is visible.

For `LogAndApply`, since DB mutex is always acquired before `ProcessManifestWrite`, even if it happens concurrently with the recovery thread, as long as it happens after `VersionSet::Recover`, the mutex establishes a happens before relationship between the recovery thread and the write thread, so `wals_` is still visible to the write thread.

Reply:
> Yes, during recovery, `GetSortedWalFiles` may need to access `wals_` concurrently with `LogAndApply`.
No, `GetSortedWalFiles` cannot happen concurrently with `Recover()`. When a thread is executing `Recover()`, no `db` ptr has been returned to the application yet. How can the application call `DB::GetSortedWal()` without a db ptr?
`GetSortedWals` is a `DB`-level API and can happen concurrently with methods that call `LogAndApply` except `Recover()`.
> 
> For `GetSortedWalFiles`, since it's in the same recovery thread, the `wals_` written during manifest recovery is visible.
> 
It's not necessarily in the same thread. It's an API, and application can call it anywhere as long as the application has the `db` ptr after recovery.

> For `LogAndApply`, since DB mutex is always acquired before `ProcessManifestWrite`, even if it happens concurrently with the recovery thread, as long as it happens after `VersionSet::Recover`, the mutex establishes a happens before relationship between the recovery thread and the write thread, so `wals_` is still visible to the write thread.

Threads performing db recovery during db open cannot happen concurrently with application write thread.

Reply:
I thought the `GetSortedWalFiles` you mentioned refers to getting the sorted WALs in `DB::Recover`.
`WalSet::CheckWals` is only called in `DB::Recover` after recovering from MANIFEST and before recovering from WAL.
`GetSortedWalFiles` public API won't verify the tracked WALs, the WAL verification only happens during `Recovery`.

Reply:
`Threads performing db recovery during db open cannot happen concurrently with application write thread.`

Yes, here I'm referring to the `LogAndApply`s called during `Recover`.